src/parser.rs

```diff
@@ -59,106 +104,528 @@ impl Header {
-#[deprecated(note = "use Someip::from_slice instead")]
-pub fn someip_message(input: &[u8]) -> Result<Message, Error> {
-    Message::from_slice(input)
-}
+impl SdPayload {
+    pub fn from_slice(s: &[u8]) -> Result<SdPayload, Error> {
+        if s.len() < SD_HEADER_LENGTH {
+            return Err(Error::NotEnoughData {
+                min: SD_HEADER_LENGTH,
+                actual: s.len(),
+            });
+        }
+        let mut buf = Cursor::new(s);
+        SdPayload::from_reader(&mut buf)
+    }
 
-impl<'a> Message<'a> {
-    /// Parse a SomeIP message from a byte array
-    pub fn from_slice(input: &'a [u8]) -> Result<Message<'a>, Error> {
-        match Header::from_slice(input) {
-            Ok(Header {
-                message_id:
-                    MessageId {
-                        service_id: 0xFFFF,
-                        method_id: 0x0000,
-                    },
-                length: 8,
-                request_id:
-                    RequestId {
-                        client_id: 0xDEAD,
-                        session_id: 0xBEEF,
-                    },
-                protocol_version: 0x01,
-                interface_version: 0x01,
-                message_type: MessageType::RequestNoReturn,
-                return_code: ReturnCode::Ok,
-            }) => Ok(Message::MagicCookieClient),
+    pub fn from_reader<R: Read>(reader: &mut R) -> Result<SdPayload, Error> {
+        let flags = reader.read_u8()?;
+        reader.read_u24::<BE>()?; // reserved
 
-            Ok(Header {
-                message_id:
-                    MessageId {
-                        service_id: 0xFFFF,
-                        method_id: 0x8000,
-                    },
-                length: 8,
-                request_id:
-                    RequestId {
-                        client_id: 0xDEAD,
-                        session_id: 0xBEEF,
-                    },
-                protocol_version: 0x01,
-                interface_version: 0x01,
-                message_type: MessageType::Notification,
-                return_code: ReturnCode::Ok,
-            }) => Ok(Message::MagicCookieServer),
+        let mut entries: Vec<SdEntry> = Vec::new();
+        let mut read: usize = 0;
+        let length: usize = match reader.read_u32::<BE>() {
+            Ok(value) => value as usize,
+            Err(_) => 0usize,
+        };
 
-            Ok(header) => {
-                let payload_len = header.length as usize - 8; // PRS_SOMEIP_00042
-                if input.len() < HEADER_LENGTH + payload_len {
-                    Err(Error::NotEnoughData {
-                        min: HEADER_LENGTH + payload_len,
-                        actual: input.len(),
-                    })
-                } else {
-                    let payload = &input[HEADER_LENGTH..HEADER_LENGTH + payload_len];
-                    Ok(Message::Message(header, payload))
+        while read < length {
```

Comment:
The nasty mut `read` binding can be avoided because the number of expected entries is known: `length / SD_ENTRY_LENGTH` e.g:

```rust
let num_entries = length / SD_ENTRY_LENGTH
for _ in num_entries { ... }
```